### PR TITLE
Internal field index rotation

### PR DESCRIPTION
**Pull request by marshallward**
This patch introduces an index rotation feature applied to all internal
fields.  When this feature is enabled, model fields are initialized on
the index map as specified by the configuration or input files.  The
values of these arrays are then moved to a rotated index map and saved
to a new rotated array, which are then used for model simulation.

The primary purpose of this feature is to detect regressions in
rotationally symmetric calculations, such as fluid dynamic dynamic
processes and other transport phenomena.

Fields created or read by the drivers are unrotated and are handled as
external user-defined field.  External fields are rotated or unrotated
as they pass between the driver and the ocean model.

In order to maintain bit reproducibility, all field initialization is
computed or assigned on the input grid.

This patch also includes an extension of the test suite, which applies a
single quarter-turn to each of the existing tests.

The features of this patch are summarized below.

Two new configuration features are added:

- ROTATE_INDEX: Boolean flag which enables this feature.
- INDEX_TURNS: Integer which descibes the number of counterclockwise
    quarter-turns.  Default value is 1.

Two new modules and several new functions are added:

- MOM_array_transform:  Fundamental operations for rotating arrays,
    vectors, rotationally equivalent pairs

- MOM_transform_FMS: Wrappers to FMS operations which can only operate
    on the unrotated index pax.

- Numerous rotation functions have been introduced in various modules
  associated with initialization (forcing, OBC, etc), which are
  documented in the source code.

The following new features have been added to existing components:

- paired checksums (`uvchksum`, `hchksum_pair`) now include a `scalar_pair`
  argument to distinguish between rotationally equivalent scalars (e.g.
  grid areas on u- and v-points) and vector components.

- `register_restart_pair` was introduced to register rotationally
  associated pairs.

- `turns` has been added to the horizontal index (HI) type, as a means
  to track the number of quarter turns in lower level operations.

- `allocate_forcing_type` and `allocate_mech_forcing` have been extended
  to permit allocation based on an exisiting (unrotated) reference,
  alongside the older interface using control flags for each field type.

- copy_MOM_domain includes a `turns` argument, to support rotational
  changes to the FMS domain.

Reproducibility:

- All results should be bitwise reproducible.  A nonphysical exception
  exception is the calculation of CFL conditions in systems with zero
  velocity, such as the unit test case.  The previous implementation
  could produce a "negative zero" CFL, whose sign would depend on the
  number of rotations.

  The new implementation eliminates these negative zeros, but may also
  change existing CFL calculations, and may report a regression in
  `ocean.stats`.

- Rotated diagnostic checksum output is not tested, since there is not
  yet a way to associated rotationally equivalent diagnostics.  See
  below for more information.

Current limitations:

- Only single PE runs can be rotated.

    Currently FMS assumes a specific ordering for assignment of PE IDs
    (east->west, south->north).  An index rotation will typically
    re-orient these domains in a way which breaks this default
    assignment.  This ID is used to manage MPI message passing
    targets, so any rearrangement of these IDs will cause errors.

    Since we currently cannot control the PE ID assigned by FMS, we are
    currently unable to support index rotation of parallel runs.

- Certain fields and tracers cannot be rotated.

    There are a small subset of tracers and fields in various
    structures, such as boundary fluxes, which are currently
    unsupported.  This is primarily beacuse these tracers are typically
    managed at the driver level.

    There is no reason why these fields cannot be supported, but it
    would require a larger test suite to detect and handle such fields.

- Diagnostic output is not rotated.

    Diagnostics are computed on the rotated grid, but it would require
    additional FMS wrappers to support registering and writing
    diagnostics on the input index map.  This is not yet implemented,
    so all diagnostics are on the rotated index map.

    One additional consequence is that there is no way to register pairs
    of diagnostics, and thus no way to automate the chksum.diag output.
    Currently, diagnostic checksums are disabled in the test suite.

    We hope to resolve this issue in a future patch.

- Open boundary condition (OBC) rotation is restriced to INDEX_TURNS=1

    Generalized OBC rotation is currently not supported.  This is
    primarily due to the sensitivity of segment configuration, which is
    specified with a large number of flags denoting its orientation.

    It is possible to support generalized rotation, but it will require
    a refactor of the OBC segment code.

    We hope to support generalized rotation in a future patch.

- Incorrect definitions of gridLonT and gridLatT

    Due to certain array length assumptions for 1d axes, we rotate and
    exchange the values of gridLatT and gridLonT during quarter-turn
    operations.

    This is not correct, since these arrays describe latitude and
    longitude, rather than the axes in the first and second dimension.
    However, several other components currently expect the first and
    second reference axis to have matching array lengths, so for now we
    swap these arrays during rotations.

    The 1d arrays are only used for initialization computation (which is
    on the input map) and I/O management, so this error does not affect
    any existing runs.  But rotated runs which save these arrays will be
    incorrect.

- Non-ALE sponge rotation is disabled.

    Only ALE sponge rotation is currently supported.

    It should be possible to extend rotation to include non-ALE sponges
    if needed.

Further comments:

- Fields are not rotated in-place, since many calculations require that
  both fields exist at any time.  While it is unlikely that rotated and
  unrotated fields would persist for an entire run, there is likely to
  be an increase in memory overhead.

- External fields from the drivers are rotated and unrotated at the
  beginning and end of every timestep, which can create additional
  computational overhead.

  Note that there are control flags, e.g. `cycle_start` and `cycle_end`
  which could be used to manage these operations, assuming that the
  driver does not change the fields inbetween coupler steps.

- This patch fixes what appears to have been an error in the global
  h-point indexes of OBC segments.  However, it also seems that these
  indices were unused, so this is unlikely to impact existing runs.

This implementation was proposed by Robert Hallberg and follows an
earlier implementation by Nic Hannah, whose efforts introduced several
changes that implemented the rotational invariance of existing numerical
calculations.

### CODE REVIEW

**Comment by marshallward** [comment]
Gaea regression test: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/pipelines/10178

File: src/framework/MOM_array_transform.F90 [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+
+  m = size(A_in, 1)
+  n = size(A_in, 2)
+
+  select case (modulo(turns, 4))
+    case(0)
+      A = A_in
+    case(1)
+      A = transpose(A_in)
+      A = A(n:1:-1, :)
```
Comment (Hallberg-NOAA) [inline src/framework/MOM_array_transform.F90:89]: The use of array syntax is generally discouraged in MOM6, as described in the MOM6 style guide (see https://github.com/NOAA-GFDL/MOM6/wiki/Code-style-guide#array-syntax). In this case, there is a certain compactness to this copy as it avoids the need for an explicit temporary array, but I think that it would be preferable to write this line as A(:,:) = A(n:1:-1, :) to make it more explicit that array syntax is being used.

File: src/framework/MOM_array_transform.F90 [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+  real, intent(out) :: A(:,:)   !< Rotated array
+
+  integer :: m, n
+
+  m = size(A_in, 1)
+  n = size(A_in, 2)
+
+  select case (modulo(turns, 4))
+    case(0)
+      A = A_in
```
Comment (Hallberg-NOAA) [inline src/framework/MOM_array_transform.F90:86]: This should be rewritten as A(:,:) = A_in(:,:) to make it more explicit what is going on.  Please see the comment a few lines below for an expanded discussion of this point.

Comment (marshallward) [reply src/framework/MOM_array_transform.F90:86]: Agreed, I will overhaul this entire file.

File: src/framework/MOM_array_transform.F90 (hunk 1 above)
Comment (adcroft) [reply src/framework/MOM_array_transform.F90:89]: Agreed. Adding `A(:,:) = A(n:1:-1,:)` works better than adding `HI` to these interfaces and since these operations are over the entire array the syntax works well. It's the partial loop ranges that are indecipherable using array syntax.

File: src/framework/MOM_transform_FMS.F90 [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+  real, optional, intent(in) :: mask_val      !> FMS mask value
+  integer, optional, intent(in) :: turns      !> Number of quarter-turns
+  integer :: chksum                           !> FMS checksum of field
+
+  real, allocatable :: field_rot(:,:)
+  integer :: qturns
+
+  qturns = 0
+  if (present(turns)) &
+    qturns = turns
```
Comment (Hallberg-NOAA) [inline src/framework/MOM_transform_FMS.F90:102]: Perhaps this would be even more elegant and robust to user silliness if this line were to read `qturns = modulo(turns, 4)`.

File: src/core/MOM_forcing_type.F90 [hunk 4]
```diff
@@ -2211,8 +2227,22 @@ subroutine mech_forcing_diags(forces, dt, G, time_end, diag, handles)
 
   integer :: i,j,is,ie,js,je
 
+  type(mech_forcing), pointer :: forces
+  integer :: turns
+
   call cpu_clock_begin(handles%id_clock_forcing)
 
+  ! NOTE: post_data expects data to be on the input index map, so any rotations
```
Comment (marshallward) [inline src/core/MOM_forcing_type.F90:2235]: This should have said "post_data expect data to be on the *rotated* index map...

File: src/framework/MOM_array_transform.F90 [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+subroutine rotate_array_pair_real_3d(A_in, B_in, turns, A, B)
+  real, intent(in) :: A_in(:,:,:)   !< Unrotated scalar array pair
+  real, intent(in) :: B_in(:,:,:)   !< Unrotated scalar array pair
+  integer, intent(in) :: turns      !< Number of quarter turns
+  real, intent(out) :: A(:,:,:)     !< Rotated scalar array pair
+  real, intent(out) :: B(:,:,:)     !< Rotated scalar array pair
+
+  integer :: k
+
+  do k = lbound(A_in, 3), ubound(A_in, 3)
```
Comment (marshallward) [inline src/framework/MOM_array_transform.F90:204]: Since these dummy arrays are 1-indexed by default, use of `lbound` and `ubound` is not adding any value and potentially causing confusion, so I will rewrite this and the other loops as 1-to-n.

File: src/framework/MOM_transform_FMS.F90 (hunk 3 above)
Comment (marshallward) [reply src/framework/MOM_transform_FMS.F90:102]: I can add it, but I will say that values outside 0-3 are handled by `rotate_array` and `allocate_rotated_array`, so it's probably not necessary in this case.

In fact, I think I have too many `modulo()` calls in the higher level functions, and would ideally only like to handle the modulo at the lowest level.

But happy to add it if you think it will help.

Comment (Hallberg-NOAA) [reply src/framework/MOM_transform_FMS.F90:102]: It probably doesn't matter too much either way since this is not performance critical code, but since there is an extra allocate and a set of copies, the modulo is probably a cheaper alternative and it would avoid unnecessarily bloating the memory of the model (which might actually matter).  However, if there is already a modulo on the number of turns in the calling code above this, then it does not help.  I would defer to your judgement, @marshallward, on the right approach here.

Comment (marshallward) [reply src/framework/MOM_transform_FMS.F90:102]: Yes, you are right.. I missed that `qturns = 4` will do the allocate.  Yeah, in that case I will apply the change.

**Comment by marshallward** [comment]
I've pushed an updated with your suggested changes, as well as my own described above, thanks @Hallberg-NOAA. 

I agree that vardesc still seems to have some value.  I personally prefer passing structs of metadata over functions with lots of arguments, but can defer this discussion.

Thanks also for indulging my whitespace indentations.  Let's schedule a whitespace battle royale at some point.

**Comment by marshallward** [comment]
Updated Gaea regression test: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/pipelines/10188

**Comment by Hallberg-NOAA** [comment]
Well, I managed to proactively mess this up!  I already updated dev/gfdl
for both MOM6 and MOM6-examples after I saw your approval, and I pushed it
before I saw the new small PR to disentangle this.  Sorry about that.

  - Bob Hallberg

 Dr. Robert Hallberg
 Oceanographer
 NOAA Geophysical Fluid Dynamics Laboratory

NOAA GFDL                             || Phone: (609) 452-6508
Princeton University Forrestal Campus || Cell:  (732) 599-0459
201 Forrestal Road                    || Fax:   (609) 987-5063
Princeton, New Jersey 08540-6649      || Email: Robert.Hallberg@noaa.gov

**Comment by adcroft** [comment]
Indeed, there are now conflicts that I had set things up to handle more smoothly. Misbehave again and I'll remove your push privileges. 😁

FYI, you did not record the MOM6 commits in the MOM6-examples commit message. I'll add them when I next update MOM6-examples.

**Comment by nichannah** [comment]
Well done Marshall!

**Comment by MJHarrison-GFDL** [comment]
Indeed!
